Run tests for Users API

tmatsuo opened the pull request:
Fixes #60 
Also adding unit tests for Users API and Modules API

bshaffer commented on .travis.yml:
```diff
@@ -73,10 +73,9 @@ script:
-  - if [ ${TRAVIS_PHP_VERSION} == '5.5' ]; then env PHP_CGI_PATH=`which php-cgi` LOCAL_TEST_TARGETS='app.yaml backend.yaml' phpunit; fi;
+  # run appengine tests
+  - pushd appengine/standard
+  - if [ ${TRAVIS_PHP_VERSION} == '5.5' ]; then testing/run_ae_e2e_tests.sh; fi;
```
Won't this fail because you're in `appengine/standard` and not the root dir? Seems like the `pushd` and `popd` should be removed here, and the one in `testing_run_ae_e2e` should be the full path.

Also, aren't these dev_appserver tests? so the name of the script should be `testing_run_dev_appserver_tests.sh`?

tmatsuo replied:
@bshaffer You're right! It failed on travis :)

Those tests include tests DeployTests too. Potentially we can skip DeployTests on travis.

tmatsuo commented:
It really took me a long time to solve, but finally I figured it out. We hit https://github.com/symfony/symfony/issues/5030#issuecomment-20678567 and the second dev_appserver didn't start. Now we have a workaround, adding `exec` before the `dev_appserver.py`, and the tests should be green!

@bshaffer PTAL

cedricziel commented on testing/GaeApp.php:
```diff
@@ -136,7 +136,8 @@ public function run(
         $phpCgiPath = '/usr/bin/php-cgi'
     ) {
-        $cmd = 'dev_appserver.py --port ' . $this->port
+        $cmd = 'exec dev_appserver.py --port ' . $this->port
```
Just curios: I've had a bad time with python 3 and dev_appserver.py in the past. - If it still depends on python2, - shouldn't we use `exec python2 dev_appserver.py` instead of relying on the env here?

tmatsuo replied:
@cedricziel Good point, thanks for raising this. That said, I'm not so concerned about this because `dev_appserver.py` should just run regardless of the system python. In other words, when Python3 is becoming norm, gcloud sdk (now `dev_appserver.py` is distributed via gcloud) should take care of it.

tmatsuo commented:
@bshaffer, rebased to the current master. PTAL

bshaffer commented:
This is really great! There's a few things I find strange:

1. E2E tests imply (to me) a `gcloud preview app deploy` command, so running `dev_appserver` is an integeration test, but not an E2E test.
2. E2E tests (which actually deploy to App Engine) should not run on travis, and only run in Jenkernetes
3. I don't like the `e2e` dir being top level. I'd rather see it be in `tests/e2e`.

Personally, I think this functionality would make much more sense if:

- The actual `e2e` tests were skipped unless an env var "RUN_E2E_TESTS" set, so it defaults to not running them
- The `dev_appserver` tests were skipped unless dev_appserver was found (this may currently be the case). Also include a `SKIP_DEV_APPSERVER_TESTS` env var that could be added in travis to skip these tests
- All tests were moved to a `test` directory, with subdirectories of `e2e` and `dev_appserver` if you prefer.

tmatsuo commented:
> E2E tests imply (to me) a gcloud preview app deploy command, so running dev_appserver is an integeration test, but not an E2E test.

I'm OK with calling devserver tests `e2e`, because it's an end to end test for local development.

> E2E tests (which actually deploy to App Engine) should not run on travis, and only run in Jenkins

Yes, I will make a change for this after this PR. I want to submit this with the green state.

> I don't like the e2e dir being top level. I'd rather see it be in tests/e2e

done

> The actual e2e tests were skipped unless an env var RUN_E2E_TESTS set, so it defaults to not running them

> The dev_appserver tests were skipped unless dev_appserver was found (this may currently be the case).

I introduced env vars for conditional run. It's kinda hard for checking dev_appserver.py is there in a portable way (until bash on windows becomes norm), so I would like to have an env var to control it.

> All tests were moved to a test directory inside the sample, with subdirectories of e2e and dev_appserver if you prefer.

Moved to `tests/e2e` subdir.

tmatsuo commented:
@bshaffer PTAL

bshaffer commented on .travis.yml:
```diff
@@ -28,6 +28,7 @@ env:
     - GOOGLE_APPLICATION_CREDENTIALS=$TRAVIS_BUILD_DIR/credentials.json
     - GOOGLE_VERSION_ID=$TRAVIS_JOB_ID
     - PATH=${HOME}/gcloud/google-cloud-sdk/bin:${PATH}
+    - RUN_DEVSERVER_TESTS=true
```
We could add this as a [matrix include](https://github.com/google/google-api-php-client/blob/master/.travis.yml#L27) only on php5.5 like this:

```yaml
matrix:
  include:
    - php: 5.5
      env: RUN_DEVSERVER_TESTS=true
```

tmatsuo replied:
Done

bshaffer commented on testing/run_ae_e2e_tests.sh:
```diff
@@ -0,0 +1,34 @@
+
+set -ex
+
+export PHP_CGI_PATH=`which php-cgi`
```
If we add this env var somewhere else (`.travis.yml` or in the test itself), we could get away with adding all these tests to `.travis.yaml` and removing this script. Then they're just going to be skipped automatically.

tmatsuo replied:
Removed!

tmatsuo commented:
Travis is sleeping, not picking up the change any more

tmatsuo commented:
Also added a unit tests.

tmatsuo commented:
Travis didn't pick it up because I had a syntax error in `.travis.yaml`! Doh

tmatsuo commented:
@bshaffer Ready for another look

bshaffer commented on .travis.yml:
```diff
@@ -17,17 +17,21 @@ branches:
+      env: RUN_DEVSERVER_TESTS=true
+    - php: 5.6
+    - php: hhvm
+    - php: nightly
```
woohoo!

bshaffer commented on .travis.yml:
```diff
@@ -17,17 +17,21 @@ branches:
     - GOOGLE_APPLICATION_CREDENTIALS=$TRAVIS_BUILD_DIR/credentials.json
     - GOOGLE_VERSION_ID=$TRAVIS_JOB_ID
     - PATH=${HOME}/gcloud/google-cloud-sdk/bin:${PATH}
+    - PHP_CGI_PATH=/home/travis/.phpenv/shims/php-cgi
```
👍

bshaffer commented on .travis.yml:
```diff
@@ -78,17 +82,22 @@ script:
+  - popd
 
-after_script:
+after_success:
```
I like this change

bshaffer commented on appengine/standard/modules/app.php:
```diff
@@ -14,6 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+namespace Google\Cloud\Samples\Appengine\Modules;
```
I haven't been namespacing `app.php`. Typically you only use `namespace` at the top of class files. What was the reasoning behind adding this?

tmatsuo replied:
This is to mock `file_get_contents`. I have `Functions.php` to mock it.

bshaffer commented on appengine/standard/modules/tests/bootstrap.php:
```diff
@@ -14,7 +14,11 @@
+require_once __DIR__ . '/../vendor/autoload.php';
 require_once __DIR__ . '/../../../../testing/E2EDeploymentTrait.php';
 require_once __DIR__ . '/../../../../testing/GaeApp.php';
 require_once __DIR__ . '/../../../../testing/LocalTestTrait.php';
```
An idea for the future (not necessarily this PR) we could include a `testing/autoload.php` which will handle registering all relevant testing files. This would be cleaner when we start adding these types of tests to other samples.

tmatsuo replied:
That's an option. Also I'm thinking to upload the testing tool to packagist, maybe with `google/cloud-testutils` or something which means it will be another independent repo. WDYT?

bshaffer replied:
Yeah, that would be even better :D

bshaffer commented on appengine/standard/modules/tests/unit/mocks/Functions.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Simple mock function for file_get_contents.
+ */
+function file_get_contents($url)
```
ahh, I think this is why you added the namespace in `app.php`?

tmatsuo replied:
Yes sir

bshaffer commented:
LGTM

tmatsuo commented:
Thanks @bshaffer ! Merging
